lnwallet/chanfunding/coin_select.go

```diff
@@ -251,3 +254,81 @@ func CoinSelectSubtractFees(feeRate chainfee.SatPerKWeight, amt,
 
 	return selectedUtxos, outputAmt, changeAmt, nil
 }
+
+// CoinSelectUpToAmount attempts to select coins such that we'll select up to
+// maxAmount exclusive of fees if sufficient funds are available. If
+// insufficient funds are available this method selects all available coins.
+func CoinSelectUpToAmount(feeRate chainfee.SatPerKWeight, maxAmount,
+	dustLimit btcutil.Amount, coins []Coin) ([]Coin, btcutil.Amount,
+	btcutil.Amount, error) {
+
+	var (
+		// Track if our coin selection was unsuccessful and whether
+		// we have to start a new round of selecting coins considering
+		// fees.
+		selectSubtractFee = false
+		outputAmount      = maxAmount
+	)
+
+	// First we try to select coins to create an output of the specified
+	// maxAmount with or without a change output that covers the miner fee.
+	selected, changeAmt, err := CoinSelect(
+		feeRate, maxAmount, dustLimit, coins,
+	)
+
+	var errInsufficientFunds *ErrInsufficientFunds
+	if errors.As(err, &errInsufficientFunds) {
+		// If the initial coin selection fails due to insufficient funds
+		// we select our total available balance minus fees.
+		selectSubtractFee = true
+	} else if err != nil {
+		return nil, 0, 0, err
+	}
+
+	// If we determined that our local balance is insufficient we check our
+	// total balance minus fees.
+	if selectSubtractFee {
+		// Get balance from coins.
+		var totalBalance btcutil.Amount
+		for _, coin := range coins {
+			totalBalance += btcutil.Amount(coin.Value)
+		}
+
+		selected, outputAmount, changeAmt, err = CoinSelectSubtractFees(
+			feeRate, totalBalance, dustLimit, coins,
+		)
+		if err != nil {
+			return nil, 0, 0, err
+		}
+	}
+
+	// Make sure that the output amount is below the specified max amount.
+	if outputAmount > maxAmount+dustLimit {
```

Comment:
question: could you explain why we are accounting for the dustLimit here. Can we not be sure that the maxAmount here is greater than the dustLimit, or should we not fail earlier in case the maxAmount is lower than the dustlimit?

Reply:
yeah, can you explain a case where this check is necessary?

Reply:
Thanks @ziggie1984 and @Crypt-iQ for following up on this.

I've had another look and am fairly convinced now that we do not need this and the following `outputAmount > maxAmount` check.

The check `outputAmount > maxAmount+dustLimit` was initially required due to an edge case in the interplay between a previous version of [`CoinSelect`](https://github.com/lightningnetwork/lnd/commit/59c40ec8b482122f732af6f0f62871beb518ab05) and `CoinSelectSubtractFees` which was described here: https://github.com/lightningnetwork/lnd/pull/4029#issuecomment-592955663.
In summary the condition could have evaluated to `true` like this: `CoinSelect` had a static fee calculation for one input and one change output, regardless of whether our available balance was sufficient to cover `maxAmount` without a change output. If we assumed that our available wallet balance was `dustLimit + 1`  satoshis greater than the sum of `maxAmount` and a fee for one input and NO change output, then `CoinSelect` would have returned `ErrInsufficientFunds` in cases where the calculated fee was greater than `dustlimit + 1`. Then `CoinSelectSubtractFees` would have taken over, selecting the total available balance( `maxAmount + dustLimit + 1` satoshis) as one output without change. Hence the `outputAmount` would have been `dustLimit + 1` satoshis greater than `maxAmount` which would have interrupted the flow at this condition.

The amended version of [`CoinSelect`](https://github.com/lightningnetwork/lnd/commit/59c40ec8b482122f732af6f0f62871beb518ab05) however now takes inputs without change outputs into account. For the above scenario this results in a UTXO selection that covers `maxAmount` without a change output and `dustLimit + 1` going towards fees. Hence we only land in `CoinSelectSubtractFees` if our wallet balance is less than `maxAmount` which imo obsoletes the `outputAmount > maxAmount` checks.

Once again appreciate your thorough review here since I find this part tricky.